### PR TITLE
INTERN-2: Peft refactor

### PR DESCRIPTION
**Pull request by Martakor1**
Made hook refactor. Now hooks creates without parameters, or with them - depends on hook child class. Parameters of hook comes from BaseCompressionModel children classes. All linkage with trainer in the trainer link_to_trainer (BaseNeuralModel).
Also there is no phase "finetune" in BasePruner, and new parameter "prune_each" in config, that allows to control pruning during training flow. There is no "ft_params" - useless.
Also lots of minor refactoring.

### CODE REVIEW

File: fedcore/models/network_impl/base_nn_model.py [hunk 1]
```diff
@@ -64,27 +65,20 @@ def __init__(self, params: Optional[OperationParameters] = None, additional_hook
         self.epochs = self.params.get("epochs", 1)
         self.batch_size = self.params.get("batch_size", 16)
         self.learning_rate = self.params.get("learning_rate", 0.001)
-        self.model = self.params.get("model", None)
-        self._init_custom_criterions(
-            self.params.get("custom_criterions", {}))  # let it be dict[name : coef], let nodes add it to trainer
```
Comment (leostre) [inline fedcore/models/network_impl/base_nn_model.py]: либо убран ценный  комментарий, либо теперь не хватает пояснения, как задавать члены лоссов

File: fedcore/algorithm/pruning/hooks.py [hunk 2]
```diff
@@ -149,48 +122,51 @@ def prune_after_reg(self, pruner, pruning_iter):
...
-    def action(self, callback_type, kws):
-        pruner_metadata = kws['pruner_objects']
-        pruner = pruner_metadata['pruner_cls']
-        pruning_iter = pruner_metadata['pruning_iterations']
-        val_dataloader = pruner_metadata['input_data'].features.val_dataloader
-        pruner = self.regularize_model_params(pruner, val_dataloader)
-        self.prune_after_reg(pruner, pruning_iter)
+    def action(self, epoch, kws):
+        pruning_iter = self.pruning_iterations
+        val_dataloader = kws['val_loader']
```
Comment (leostre) [inline fedcore/algorithm/pruning/hooks.py]: я упоминал тогда вечером, что использование валидационного датасета при прунинге приводит к data leakage и необходимо выделение некоторого подмножества из трейна

**Review by leostre — APPROVED** [review]
Слушай, в целом, адекватно, тесты тоже достаточно логично и явно выглядят, раз проходят, видимо, хорошо. Единственное, меня смущает логика link_to_trainer и hooks  extesion относительно модели BaseNeural. Было бы славно убедиться, что вызов например, с prune_each & save_each приводит к инстанцированию как ZeroShotPruner, так и Saver

File: fedcore/models/network_impl/base_nn_model.py (hunk 1 above)
Comment (Martakor1) [reply fedcore/models/network_impl/base_nn_model.py]: Добавлю тогда в NeuralModelConfigTemplate, мне кажется в месте объявления этого параметра комментарий будет смотреться лучше.

File: fedcore/algorithm/pruning/hooks.py [hunk 3]
```diff
@@ -1,144 +1,117 @@
...
+import torch_pruning as tp
+if TYPE_CHECKING:
+    from fedcore.models.network_impl.base_nn_model import BaseNeuralModel
+    
 
 
 class ZeroShotPruner(BaseHook):
-    _SUMMON_KEY = 'pruning'
+    _SUMMON_KEY = 'prune_each'
     _hook_place = 50
```
[the remark this answers is not in the document]
Comment (Martakor1) [reply fedcore/algorithm/pruning/hooks.py]: Поля pruning я не наблюдал и оно не работало

File: fedcore/models/network_impl/base_nn_model.py [hunk 4]
```diff
@@ -52,9 +51,11 @@ class BaseNeuralModel(torch.nn.Module):
                 print(features)
     """
 
-    def __init__(self, params: Optional[OperationParameters] = None, additional_hooks=None):
+    BASE_HOOKS : list[type['BaseHook']] = LOGGING_HOOKS + MODEL_LEARNING_HOOKS
+
+    def __init__(self, model: torch.nn.Module, params: dict = {}, additional_hooks: Sequence['BaseHook'] = []):
```
[the remark this answers is not in the document]
Comment (Martakor1) [reply fedcore/models/network_impl/base_nn_model.py]: Почему логичнее? Кажется, BaseNeuralModel нет смысла существовать без какой-то внутренней имплементации. Модель - неотъемлемая часть этого класса.

File: fedcore/algorithm/low_rank/svd_tools.py [hunk 5]
```diff
@@ -17,11 +17,11 @@
...
 __all__ = [
-    'decompose_module',
+    'decompose_module_in_place',
     'load_svd_state_dict'
 ]
 
-def decompose_module(
+def decompose_module_in_place(
     model: Module,
     decomposing_mode: Optional[str] = True,
```
Comment (Martakor1) [inline fedcore/algorithm/low_rank/svd_tools.py]: Абсолютно не понятно, как может Optional[str] = True. Отслеживал по коду, сначала была строка, потом стал бул. Не ясна семантика поля.

File: fedcore/algorithm/pruning/hooks.py [hunk 6]
```diff
@@ -342,8 +314,19 @@ def eval_entropy():
...
+    callback = ZeroShotPruner
+    if pruner_in_depth:
+        callback = PrunerInDepth
+    elif pruner_with_reg:
+        callback = PrunerWithReg
+    elif pruner_with_grads:
+        callback = PrunerWithGrad
+    elif zeroshot_pruner:
+        callback = ZeroShotPruner
+    return callback
```
Comment (Martakor1) [inline fedcore/algorithm/pruning/hooks.py]: Так и не понял, что тут хотели. Очень странно отбираются прунеры. Есть пересечения в названиях и остальном. Кажется, что поломано.

File: fedcore/models/network_impl/hooks.py
Comment (Martakor1) [inline fedcore/models/network_impl/hooks.py]: Evaluator: inputs = tuple(inputs_.to(self.device) for inputs_ in inputs if hasattr(inputs_, 'to')) - разве могут быть не torch.Tensor без to?

У Freezer поле criterion означает не совсем тот criterion, который loss, может есть другое название, более подходящее?

**Comment by Martakor1** [comment]
В целом я оставил все вопросы, которые накопил по коду.

File: fedcore/models/network_impl/base_nn_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leostre) [reply fedcore/models/network_impl/base_nn_model.py]: это да, но часто у нас может быть создан пайплайн, а вот модель появляться только на стадии fit, потому что является частью входных данных.
Подобная логика происходит со структурой данных в sklearn, например

File: fedcore/algorithm/pruning/hooks.py [hunk 7]
```diff
@@ -1,144 +1,112 @@
...
+        if isinstance(self.criterion_for_pruner, torch.nn.CrossEntropyLoss):  # classification task
             target = target.to(torch.int64)  # convert probalistic output to labels)
         try:
-            loss = self.criterion_for_grad(out, target)
+            loss = self.criterion_for_pruner(out, target)
         except Exception:
-            loss = self.criterion_for_grad(out, target)
+            loss = self.criterion_for_pruner(out, target)
         loss.backward()
         return loss
```
[the remark this answers is not in the document]
Comment (Martakor1) [reply fedcore/algorithm/pruning/hooks.py]: Я про вот это (try и except делают одно и то же)
        
```try:
            loss = self.criterion_for_grad(out, target)
        except Exception:
            loss = self.criterion_for_grad(out, target)
        loss.backward()
        return loss

File: fedcore/models/network_impl/base_nn_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Martakor1) [reply fedcore/models/network_impl/base_nn_model.py]: Может бы речь идёт про LowRank, BasePruner и так далее? У них осталась возможность создаваться без model. И действительно в них модель приходит на .fit(). Но для BaseNeuralModel я посмотрел - везде он инициализируется уже когда есть model, так как этот класс обычно выполняет функцию внутреннего trainer - то есть обёртки над моделью. Пока не вижу прямого использования этого класса без model.
Но если может появляться BaseNeuralModel где-то в пайплайне как самостоятельная модель, то окей, переделаю. Правда нет особо в этом смысле, потому что что она будет делать тогда, только базовые хуки использовать? То есть просто обучение пустое без ничего?

Кстати, пока что там итак стоит в BaseNM.fit() строка self.model = input_data.target if self.model is None else self.model. Мб просто пометить, что параметр это Optional[] ?

File: fedcore/algorithm/low_rank/svd_tools.py (hunk 5 above)
Comment (Martakor1) [reply fedcore/algorithm/low_rank/svd_tools.py]: Сделал

**Comment by Martakor1** [comment]
Коротко, что произошло: (коротко не получилось)
Я отребейзился на merge commit https://github.com/v1docq/FedCore/pull/42/changes
Комиты ребейза идут до upd: Add DecomposerType[Enum] [ab04f6c](https://github.com/v1docq/FedCore/pull/37/commits/ab04f6c6cf89a85e2cba396c75ec5f61651e6f80), после этого идут новые коммиты фиксы.
Что фиксилось глобально?
Учитывались старые, но изменённые _init в Quantizer, LowRank и Pruner и BaseNM и Base_Compr_Model. Постарался перенести логику с registry и выбором трейнера.
Теперь создание трейнера более универсально, но требуется проверка работы registry и идеи с копированием чекпоинтом и тд.
!Логика с registry видимо полетела, но ещё до меня, так как в исходной ветке (https://github.com/v1docq/FedCore/pull/33/changes?new_files_changed=true#top) было больше смысла , чем в мёрженой. Оставил комменты в TODO, так как я не могу отвалидировать пока эту логику.
Убрана излишняя документация
Починено много импортов хуков
Чуть-чуть переработаны IDecomposed модули и логика передачи в них decomposer упрощена
Поддержаны новые api_configs и LearningConfigTemplate
После мёржа поломались места с input_data.target - заменил на input_data.features.target во многих местах (Почему-то там теперь InputData а не CompressionInputData)
Убран лишний _get_example_batch в LowRank так как в бейз классе есть такой же
Исправлены тесты, теперь нет поддержки поля peft_strategy в  LearningConfigTemplate, peft_task передаётся в data_check вручную, не через конфиг.
Убраны импорты и оптимизирована логика работы с logger
Убран во второй раз (так как после ребейза он опять появился) finetune в pruner, так как с новыми хуками мы решили, что этой секции нет как отдельной, а регулируется через prune_each сколько steps будет до и после прунинга.

Протестировано, но нуждается в доработках (фиксирую состояние на этот момент):
Запускались базовые тесты low_rank, quantizer и pruner. Классы отрабатывают. Однако в quantizer теперь ошибка для static и qat, что-то не так с кернелами для CPU в методе summary() где quantizer.estimate_params(). Хотя раньше вроде работало.
2 теста отваливаются в transmla для reassembler (но это не касается хуков)
2 теста отваливаются в test_quant_utils, но они тоже на QDQWrapper
1 тест test_linear2 отваливается в test_layers - почему то RSVD декомпозер не может разложить матрицу W (какая-то ошибка с шейпами при matmul внутри)
Помимо этого, возможно осталось ещё много тестов, но не запущенных, так как всё ещё остаются ошибки с imports и кодом, который уже меньше касается хуков и базовых классов, а касается LLM экспериментов Иры.

Итог: авторам коммитов, попавших в мёрж нужно проверить свою логику. А именно: registry и сохранение моделей. LLM тесты и хуки на них. + желательно бы дофиксить тесты.

Дополнительные комментарии: 1) неопределённость InputData и CompressionData создаёт всё больше запутанностей в коде, как создала сейчас. 2) Избыточная автогенеренная документация выпилена/сокращена, но не везде, желательно довыпиливать, так как файлы раздуваются x2 в размере.

Предложения: лучше как можно скорее влить эту ветку, пусть и не полностью доработанную, иначе поддерживать разные версии кода хуков и прочего будет не реально, так как эта логика является довольно центральной в коде. Допилить тесты или что-то конкретное локально уже смогут авторы изменений.
